libraries/botbuilder-dialogs/src/memory/scopes/settingsMemoryScope.ts

```diff
@@ -25,16 +35,110 @@ export class SettingsMemoryScope extends MemoryScope {
-            if (typeof process.env[key] == 'string') {
-                settings[key] = process.env[key];
+        } else {
+            const configuration = dc.context.turnState.get(DialogTurnStateConstants.configuration);
+            if (configuration) {
+                settings = SettingsMemoryScope.loadSettings(configuration);
+                dc.context.turnState.set(ScopePath.settings, settings);
             }
         }
+        const envSettings = SettingsMemoryScope.loadSettings(process.env);
+        merge(settings, envSettings);
         return settings;
     }
+
+    /**
+     * Build a dictionary view of configuration providers.
+     *
+     * @param configuration The configuration that we are running with.
+     * @returns Projected dictionary for settings.
+     */
+    protected static loadSettings(configuration: Record<string, string>): Record<string, unknown> {
+        const settings = {};
+        if (configuration) {
+            // load configuration into settings
+            const root = this.convertFlattenSettingToNode(Object.entries(configuration));
+            root.children.forEach((u) => {
+                settings[u.value] = this.convertNodeToObject(u);
+            });
+        }
+        return settings;
+    }
+
+    private static convertFlattenSettingToNode(kvs: [string, string][]): Node {
+        const root = new Node();
+
+        kvs.forEach(([key, value]) => {
+            const keyChain = key.split(':');
+            let currentNode = root;
+            keyChain.forEach((item) => {
+                const matchItem = currentNode.children.find((u) => u?.value === item);
+                if (!matchItem) {
+                    // Remove all the leaf children
+                    currentNode.children = currentNode.children.filter((u) => u.children.length !== 0);
+
+                    // Append new child into current node
+                    const node = new Node(item);
+                    currentNode.children.push(node);
+                    currentNode = node;
+                } else {
+                    currentNode = matchItem;
+                }
+            });
+
+            currentNode.children.push(new Node(value));
+        });
+
+        return root;
+    }
+
+    private static convertNodeToObject(node: Node): unknown {
+        if (!node.children.length) {
+            return {};
+        }
+
+        // If the child is leaf node, return its value directly.
+        if (node.children.length === 1 && node.children[0].children.length === 0) {
+            return node.children[0].value;
+        }
+
+        // check if all the children are number format.
+        let pureNumberIndex = true;
+        let indexArray: number[] = [];
+        let indexMax = -1;
+        for (let i = 0; i < node.children.length; i++) {
+            const child = node.children[i];
+            if (/^-?\d+$/.test(child.value)) {
+                const num = parseInt(child.value);
```

Comment:
`parseInt` should take a `base` arg. Also, why not just use `parseInt` as the outer check in place of the regular expression?

Reply:
`parseInt` would parse float strings into integer numbers with no error, so it cannot be used to check if a string is an integer.